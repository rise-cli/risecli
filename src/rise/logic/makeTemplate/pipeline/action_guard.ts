type GuardInput = {
    path?: string
    field: string
    index: number
    config: {
        pk?: string
        pk2?: string
        pk3?: string
        sk: string
    }
}

export function setKey(value: string) {
    const path = '$ctx.args.input.'
    const cogPath = '$ctx.identity.claims.'

    if (value?.startsWith('$')) {
        return value.replace('$', path)
    } else if (value?.startsWith('!')) {
        return value.replace('!', cogPath)
    } else {
        return `"${value}"`
    }
}

export function makeGuardFunction(props: GuardInput) {
    let pkMode = 'pk'
    let partitionKey = 'pk'

    if (props.config.pk) {
        pkMode = 'pk'
        partitionKey = setKey(props.config.pk)
    }

    if (props.config.pk2) {
        pkMode = 'pk2'
        partitionKey = setKey(props.config.pk2)
    }

    if (props.config.pk3) {
        pkMode = 'pk3'
        partitionKey = setKey(props.config.pk3)
    }

    let sk = setKey(props.config.sk)

    return {
        Resources: {
            [`FunctionGaurd${props.field}${props.index}`]: {
                Type: 'AWS::AppSync::FunctionConfiguration',
                Properties: {
                    ApiId: {
                        'Fn::GetAtt': ['GraphQlApi', 'ApiId']
                    },
                    DataSourceName: 'DbDatasource',
                    FunctionVersion: '2018-05-29',
                    Name: `FunctionGaurd${props.field}${props.index}`,
                    RequestMappingTemplate: `
                {
                  "version": "2017-02-28",
                  "operation": "GetItem",
                  "key": {
                     "${pkMode}" : $util.dynamodb.toDynamoDBJson(${partitionKey}),
                     "sk" : $util.dynamodb.toDynamoDBJson(${sk})
                  },
                }`,

                    ResponseMappingTemplate: `
                    #if(!$ctx.result)
                        $util.error("Unauthorized")
                    #else
                        $util.toJson($ctx.result)
                    #end    
                `
                },
                DependsOn: ['DbDatasource', 'GraphQlSchema']
            }
        }
    }
}
