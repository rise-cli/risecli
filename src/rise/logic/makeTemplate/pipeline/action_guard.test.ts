import { makeGuardFunction, setKey } from './action_guard'

test('setKey will return same thing if no special characters are used', () => {
    const result = setKey('note')
    expect(result).toBe('"note"')
})

test('setKey will return cognito reference if ! is used', () => {
    const result = setKey('!note')
    expect(result).toBe('$ctx.identity.claims.note')
})

test('setKey will return input reference if $ is used', () => {
    const result = setKey('$note')
    expect(result).toBe('$ctx.args.input.note')
})

test('makeGuardFunction will make CF JSON for guard vtl function', () => {
    const input = {
        field: 'getNote',
        index: 2,
        config: {
            pk: '!sub',
            sk: '$id'
        }
    }
    const result = makeGuardFunction(input)
    expect(result).toMatchSnapshot()
})

test('makeGuardFunction will target pk2 if its defined', () => {
    const input = {
        field: 'getNote',
        index: 2,
        config: {
            pk2: '!sub',
            sk: '$id'
        }
    }
    const result = makeGuardFunction(input)
    expect(result).toMatchSnapshot()
})
