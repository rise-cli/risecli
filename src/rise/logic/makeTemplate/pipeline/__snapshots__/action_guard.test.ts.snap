// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeGuardFunction will make CF JSON for guard vtl function 1`] = `
Object {
  "Resources": Object {
    "FunctionGaurdgetNote2": Object {
      "DependsOn": Array [
        "DbDatasource",
        "GraphQlSchema",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GraphQlApi",
            "ApiId",
          ],
        },
        "DataSourceName": "DbDatasource",
        "FunctionVersion": "2018-05-29",
        "Name": "FunctionGaurdgetNote2",
        "RequestMappingTemplate": "
                {
                  \\"version\\": \\"2017-02-28\\",
                  \\"operation\\": \\"GetItem\\",
                  \\"key\\": {
                     \\"pk\\" : $util.dynamodb.toDynamoDBJson($ctx.identity.claims.sub),
                     \\"sk\\" : $util.dynamodb.toDynamoDBJson($ctx.args.input.id)
                  },
                }",
        "ResponseMappingTemplate": "
                    #if(!$ctx.result)
                        $util.error(\\"Unauthorized\\")
                    #else
                        $util.toJson($ctx.result)
                    #end    
                ",
      },
      "Type": "AWS::AppSync::FunctionConfiguration",
    },
  },
}
`;

exports[`makeGuardFunction will target pk2 if its defined 1`] = `
Object {
  "Resources": Object {
    "FunctionGaurdgetNote2": Object {
      "DependsOn": Array [
        "DbDatasource",
        "GraphQlSchema",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GraphQlApi",
            "ApiId",
          ],
        },
        "DataSourceName": "DbDatasource",
        "FunctionVersion": "2018-05-29",
        "Name": "FunctionGaurdgetNote2",
        "RequestMappingTemplate": "
                {
                  \\"version\\": \\"2017-02-28\\",
                  \\"operation\\": \\"GetItem\\",
                  \\"key\\": {
                     \\"pk2\\" : $util.dynamodb.toDynamoDBJson($ctx.identity.claims.sub),
                     \\"sk\\" : $util.dynamodb.toDynamoDBJson($ctx.args.input.id)
                  },
                }",
        "ResponseMappingTemplate": "
                    #if(!$ctx.result)
                        $util.error(\\"Unauthorized\\")
                    #else
                        $util.toJson($ctx.result)
                    #end    
                ",
      },
      "Type": "AWS::AppSync::FunctionConfiguration",
    },
  },
}
`;
